File: day2/laserstrike/submissions/accepted/jb.py
#!/usr/bin/env python3

phase, n = [int(x) for x in input().split()]

#################### ALICE ####################
if phase == 1:
  g = [[] for i in range(n)]
  for i in range(n-1):
    a, b = [int(x) for x in input().split()]
    g[a].append(b)
    g[b].append(a)

  leafs = []
  leafs_hanging = [[] for i in range(n)]

  def touch(x):
    if len(g[x]) == 1:
      leafs.append(x)
      leafs_hanging[g[x][0]].append(x)

  for x in range(n):
    touch(x)

  def cut(par, leaf):
    print(leaf)
    g[par].remove(leaf)
    leafs.remove(leaf);
    touch(par)
    if len(g[par]) == 0:
      exit()

  def trim(x):
    for y in leafs_hanging[x]:
      cut(x, y)
    leafs_hanging[x].clear()

  xe = [(x,leafs_hanging[x][0]) for x in range(n) if len(leafs_hanging[x])]
  red  = sorted([(x,y) for x,y in xe if x < y])
  blue = sorted([(y,x) for x,y in xe if x > y])
  bit = len(blue) == 0 or (len(red) > 0 and red[-1] > blue[0])
  print(int(bit))

  order = red + blue if bit else blue + red
  for leaf_edge in order:
    for x in leaf_edge:
      trim(x)

  while leafs:
    trim(g[leafs[0]][0])


####################  BOB  ####################
if phase == 2:
  typ = int(input())
  seen = [0]*n
  time = 99
  last_leaf_edge = (-1,-1)
  for i in range(n-1):
    a, b = sorted([int(x) for x in input().split()])
    t = (seen[a] == time or seen[a] < seen[b] < time)
    if not seen[a] and not seen[b]:
      if (a,b) < last_leaf_edge:
        typ = not typ
      last_leaf_edge = (a,b)
      t = typ
    time += 1
    seen[a] = seen[b] = time
    print(b if t else a)

File: day2/laserstrike/submissions/accepted/jb.cc
#include <bits/stdc++.h>
using namespace std;

#define rep(i, a, b) for(int i = a; i < (b); ++i)
#define all(x) begin(x), end(x)
#define sz(x) (int)(x).size()
#define debug(...) //ignore
typedef long long ll;
typedef pair<int, int> pii;
typedef vector<int> vi;
typedef long double ld;

int main() {
  int phase, n;
  cin >> phase >> n;
  if(phase == 1) { // ALICE
    vector<vi> g(n);
    vi deg(n), xors(n);
    rep(i,1,n) {
      int x,y;
      cin>>x>>y;
      g[x].emplace_back(y);
      g[y].emplace_back(x);
      ++deg[x];
      ++deg[y];
      xors[x] ^= y;
      xors[y] ^= x;
    }

    list<int> todo;
    vector<vi> leafs_hanging(n);
    rep(x,0,n) if(deg[x] == 1)
        leafs_hanging[xors[x]].emplace_back(x);

    auto cut = [&](int par, int leaf){
      cout << leaf << endl;
      --deg[par];
      --deg[leaf];
      xors[par] ^= leaf;
      if(deg[par] == 1) {
        todo.emplace_back(par);
        leafs_hanging[xors[par]].emplace_back(par);
      }
    };

    auto trim = [&](int x){
      for(int y : leafs_hanging[x]) cut(x,y);
      leafs_hanging[x].clear();
    };

    vector<pii> red, blue;
    rep(x,0,n) if(sz(leafs_hanging[x])) {
      int y = leafs_hanging[x][0];
      if(x < y) red.emplace_back(x,y);
      else blue.emplace_back(y,x);
    }
    sort(all(red));
    sort(all(blue));
    bool bit = sz(blue) == 0 || (sz(red) > 0 && red.back() > blue[0]);
    cout << int(bit) << endl;

    if(bit) for(auto [x,y] : red) trim(x);
    for(auto [y,x] : blue) trim(x);
    if(!bit) for(auto [x,y] : red) trim(x);

    for(int x : todo) if(deg[x]) trim(xors[x]);
  }
  else { // BOB
    int typ, time = 99;
    cin>>typ;
    vi seen(n);
    pii last_leaf_edge{-1,-1};
    rep(i,1,n) {
      int a, b;
      cin >> a >> b;
      if(b < a) swap(a,b);
      bool t = (seen[a] == time || (seen[a] < seen[b] && seen[b] < time));
      if(!seen[a] && !seen[b]) {
        if(pii(a,b) < last_leaf_edge) typ = !typ;
        last_leaf_edge = pii(a,b);
        t = typ;
      }
      seen[a] = seen[b] = ++time;
      cout << (t ? b : a) << endl;
    }
  }
}

